src/node_buffer.cc

```diff
@@ -19,7 +19,11 @@
   } while (0)
 
 #define ARGS_THIS(argT)                                                     \
-  Local<Object> obj = argT;                                                 \
+  Handle<Value> _obj = argT;                                                \
+  if (!HasInstance(_obj)) {                                                 \
+    return env->ThrowTypeError("must be called on Buffer instance");        \
```

Comment:
I'm not a fan of control flow in macros and I think that the buffer code in general is overdue for an overhaul.

I want to suggest a more structured approach where the JS prototype methods typecheck the `this` argument before passing it on to their C++ counterparts.  I imagine it would look something like this (using CamelCase for dramatic effect):

```javascript
const binding = process.binding('buffer');
const BindingIsBuffer = binding.IsBuffer;
const BindingToString = binding.ToString;
const BindingUtf8Slice = binding.Utf8Slice;
// ...

// keep a reference to Object#toString, the user may monkey-patch it
const ObjectProtoToString = Object.prototype.toString;

function IsBuffer(obj) {
  return typeof(obj) === 'object' && obj !== null && BindingIsBuffer(obj);
}

// ...

Buffer.prototype.toString = function() {
  if (!IsBuffer(this))
    return ObjectProtoToString(this);

  if (this.length === 0)
    return '';

  return BindingToString(this);
};

Buffer.prototype.utf8Slice = function(start, end) {
  if (!IsBuffer(this))
    throw new TypeError('this is not an instance of Buffer');

  // validate and coerce start and end

  if (this.length === 0 && start === 0 && end === 0)
    return '';

  return BindingUtf8Slice(this, start, end);
};

// etc.
```

In short, be paranoid and rigorous.  It can be more paranoid still; for example, the user may have replaced `this.length` with a getter so it would be prudent to read it only once.